src/shims/unix/foreign_items.rs

```diff
@@ -729,6 +729,18 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
                 this.write_scalar(Scalar::from_i32(-1), dest)?;
             }
 
+            "_Unwind_RaiseException" => {
```

Comment:
So is this a standard Unix function? What are the documented guarantees for how it behaves? We should be careful to not implement it in a way that provides fewer or more guarantees than the official spec.

Reply:
It was originally specified as part of the Itanium C++ ABI afaik: https://itanium-cxx-abi.github.io/cxx-abi/abi-eh.html#base-throw On Linux it is documented as part of the LSB: https://refspecs.linuxfoundation.org/LSB_5.0.0/LSB-Core-generic/LSB-Core-generic/baselib--unwind-raiseexception.html Basically every other UNIX uses the exact same api though. Arm also references back to the Itanium C++ ABI for the definition of `_Unwind_RaiseException` for arm64: https://github.com/ARM-software/abi-aa/blob/main/cppabi64/cppabi64.rst#toc-entry-35 For arm32 they did something custom, but similar enough that the same `_Unwind_RaiseException` impl in miri should work: https://github.com/ARM-software/abi-aa/blob/main/ehabi32/ehabi32.rst

Reply:
Thanks! So this is way more standard than I thought.

This seems a bit asymmetric though, one can now start unwinding with a standard function but how does one catch it? Is it normal that that still requires a magic builtin? Hm I guess that's not even a Miri magic function, that's an intrinsic.

Final point, how does this work when different languages do unwinding? My understanding is that this is somehow distinguished, there is such a thing as "foreign unwinding" coming into Rust code which is different from a Rust panic -- but how is it different? Just different payload behind this pointer? Is that why the payload must have type `_Unwind_Exception`? Maybe Miri should check that, as [these docs](https://itanium-cxx-abi.github.io/cxx-abi/abi-eh.html#base-throw) indicate, the exception_class and exception_cleanup fields are set? And Miri assumes all unwinding is Rust unwinding so exception_class must be set to whatever the magic value is that Rust uses there.

Reply:
> The current std::panic::catch_unwind implementation does abort when catching a foreign exception by calling __rust_foreign_exception()

But nothing in Miri sets this? So with `miri_start_panic` does the caller also set this  "M O Z \0 R U S T" class?

Reply:
Oh I get it, `catch_unwind` only does that when using the "gcc" unwinding implementation. Basically the Miri unwinding "ABI" doesn't currently have support for indicating the exception class.

But this means it should be possible to throw a foreign exception with `_Unwind_Exception` and then catch it with `try` (what a terrible name^^) and observe the payload... Oh wait, that's exactly what your test is doing.^^